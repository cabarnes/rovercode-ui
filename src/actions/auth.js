// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const UPDATE_VALID_AUTH = 'UPDATE_VALID_AUTH';
export const USER_LOGOUT = 'USER_LOGOUT';

// action creators
export const updateValidAuth = (isValidAuth) => ({
  type: UPDATE_VALID_AUTH,
  payload: isValidAuth,
});

export const logout = () => ({
  type: USER_LOGOUT,
});

// helper functions
export const checkAuthError = (dispatch) => (error) => {
  if (error.response && error.response.status === 401) {
    // Authentication is no longer valid
    dispatch(updateValidAuth(false));
  } else {
    throw error;
  }
};

export const authHeader = (cookies) => ({
  headers: {
    Authorization: `JWT ${cookies.get('auth_jwt')}`,
  },
});
