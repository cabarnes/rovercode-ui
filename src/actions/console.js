// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const APPEND = 'APPEND';
export const CLEAR = 'CLEAR';

// action creators
export const append = (message) => ({
  type: APPEND,
  payload: message,
});

export const clear = () => ({
  type: CLEAR,
});
