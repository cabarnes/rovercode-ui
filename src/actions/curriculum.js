// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

import axios from 'axios';

export const FETCH_COURSES = 'FETCH_COURSES';
export const FETCH_COURSES_PENDING = `${FETCH_COURSES}_PENDING`;
export const FETCH_COURSES_FULFILLED = `${FETCH_COURSES}_FULFILLED`;
export const FETCH_COURSES_REJECTED = `${FETCH_COURSES}_REJECTED`;

// action creators
export const fetchCourses = (xhrOptions) => ({
  type: FETCH_COURSES,
  meta: {
    debounce: {
      time: 500,
    },
  },
  payload: axios.get('/api/v1/courses/', xhrOptions)
    .then(({ data }) => (
      data
    )),
});
