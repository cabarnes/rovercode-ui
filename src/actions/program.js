// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

import axios from 'axios';

export const CLEAR_PROGRAMS = 'CLEAR_PROGRAMS';
export const FETCH_PROGRAMS = 'FETCH_PROGRAMS';
export const FETCH_PROGRAMS_PENDING = `${FETCH_PROGRAMS}_PENDING`;
export const FETCH_PROGRAMS_FULFILLED = `${FETCH_PROGRAMS}_FULFILLED`;
export const FETCH_PROGRAMS_REJECTED = `${FETCH_PROGRAMS}_REJECTED`;
export const REMOVE_PROGRAM = 'REMOVE_PROGRAM';
export const REMOVE_PROGRAM_PENDING = `${REMOVE_PROGRAM}_PENDING`;
export const REMOVE_PROGRAM_FULFILLED = `${REMOVE_PROGRAM}_FULFILLED`;
export const REMOVE_PROGRAM_REJECTED = `${REMOVE_PROGRAM}_REJECTED`;

// action creators
export const clearPrograms = () => ({
  type: CLEAR_PROGRAMS,
});

export const fetchPrograms = (xhrOptions) => ({
  type: FETCH_PROGRAMS,
  meta: {
    debounce: {
      time: 500,
    },
  },
  payload: axios.get('/api/v1/block-diagrams/', xhrOptions)
    .then(({ data }) => (
      data
    )),
});

export const removeProgram = (id, xhrOptions) => ({
  type: REMOVE_PROGRAM,
  payload: axios.delete(`/api/v1/block-diagrams/${id}/`, xhrOptions)
    .then(({ data }) => (
      data
    )),
});
