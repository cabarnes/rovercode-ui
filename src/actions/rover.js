// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const SCAN = 'SCAN';
export const SCAN_FULFILLED = `${SCAN}_FULFILLED`;
export const SCAN_REJECTED = `${SCAN}_REJECTED`;
export const CONNECT_ROVER = 'CONNECT_ROVER';
export const CONNECT_ROVER_FULFILLED = `${CONNECT_ROVER}_FULFILLED`;
export const CONNECT_ROVER_REJECTED = `${CONNECT_ROVER}_REJECTED`;
export const SEND_ROVER = 'SEND_ROVER';
export const SEND_ROVER_FULFILLED = `${SEND_ROVER}_FULFILLED`;
export const SEND_ROVER_REJECTED = `${SEND_ROVER}_REJECTED`;
export const DISCONNECT_ROVER = 'DISCONNECT_ROVER';

const UART_SERVICE_UUID = '6e400001-b5a3-f393-e0a9-e50e24dcca9e';
const UART_TX_CHARACTERISTIC_UUID = '6e400002-b5a3-f393-e0a9-e50e24dcca9e';
const UART_RX_CHARACTERISTIC_UUID = '6e400003-b5a3-f393-e0a9-e50e24dcca9e';

// action creators
export const scan = () => ({
  type: SCAN,
  payload: navigator.bluetooth.requestDevice({
    filters: [{ namePrefix: 'BBC micro:bit' }],
    optionalServices: [UART_SERVICE_UUID],
  }),
});

export const connect = (device, onMessage) => ({
  type: CONNECT_ROVER,
  payload: device.gatt.connect()
    .then(server => server.getPrimaryService(UART_SERVICE_UUID))
    .then(service => Promise.all([
      service.getCharacteristic(UART_RX_CHARACTERISTIC_UUID),
      service.getCharacteristic(UART_TX_CHARACTERISTIC_UUID)
        .then((characteristic) => {
          characteristic.startNotifications();
          characteristic.addEventListener(
            'characteristicvaluechanged',
            onMessage,
          );
          return characteristic;
        }),
    ])),
});

export const send = (channel, message) => ({
  type: SEND_ROVER,
  payload: channel.writeValue(message),
});

export const disconnect = device => ({
  type: DISCONNECT_ROVER,
  payload: device.gatt.disconnect(),
});
