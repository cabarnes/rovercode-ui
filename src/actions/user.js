// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions
import axios from 'axios';

export const UPDATE_USER = 'UPDATE_USER';
export const EDIT_USER_USERNAME = 'EDIT_USER_USERNAME';
export const EDIT_USER_USERNAME_PENDING = `${EDIT_USER_USERNAME}_PENDING`;
export const EDIT_USER_USERNAME_FULFILLED = `${EDIT_USER_USERNAME}_FULFILLED`;
export const EDIT_USER_USERNAME_REJECTED = `${EDIT_USER_USERNAME}_REJECTED`;
export const EDIT_USER_PASSWORD = 'EDIT_USER_PASSWORD';
export const EDIT_USER_PASSWORD_PENDING = `${EDIT_USER_PASSWORD}_PENDING`;
export const EDIT_USER_PASSWORD_FULFILLED = `${EDIT_USER_PASSWORD}_FULFILLED`;
export const EDIT_USER_PASSWORD_REJECTED = `${EDIT_USER_PASSWORD}_REJECTED`;
export const EDIT_USER_SHOW_GUIDE = 'EDIT_USER_SHOW_GUIDE';
export const EDIT_USER_SHOW_GUIDE_PENDING = `${EDIT_USER_SHOW_GUIDE}_PENDING`;
export const EDIT_USER_SHOW_GUIDE_FULFILLED = `${EDIT_USER_SHOW_GUIDE}_FULFILLED`;
export const EDIT_USER_SHOW_GUIDE_REJECTED = `${EDIT_USER_SHOW_GUIDE}_REJECTED`;

// action creators
export const updateUser = (data) => ({
  type: UPDATE_USER,
  payload: data,
});

export const editUserUsername = (username, xhrOptions) => ({
  type: EDIT_USER_USERNAME,
  payload: axios.put('/jwt/auth/user/', { username }, xhrOptions)
    .then(({ data }) => (
      data
    )),
});

export const editUserPassword = (password, xhrOptions) => ({
  type: EDIT_USER_PASSWORD,
  payload: axios.post('/jwt/auth/password/change/', {
    new_password1: password,
    new_password2: password,
  }, xhrOptions)
    .then(({ data }) => (
      data
    )),
});

export const editUserShowGuide = (id, show, xhrOptions) => ({
  type: EDIT_USER_SHOW_GUIDE,
  payload: axios.patch(`/api/v1/users/${id}/`, { show_guide: show }, xhrOptions)
    .then(({ data }) => (
      data
    )),
});
