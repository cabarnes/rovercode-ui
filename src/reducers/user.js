import {
  UPDATE_USER,
  EDIT_USER_USERNAME_PENDING,
  EDIT_USER_USERNAME_FULFILLED,
  EDIT_USER_USERNAME_REJECTED,
  EDIT_USER_PASSWORD_PENDING,
  EDIT_USER_PASSWORD_FULFILLED,
  EDIT_USER_PASSWORD_REJECTED,
  FETCH_USER_LIST_PENDING,
  FETCH_USER_LIST_FULFILLED,
  FETCH_USER_LIST_REJECTED,
} from '../actions/user';

export default function user(
  state = {
    user_id: null,
    username: null,
    email: null,
    exp: null,
    isSocial: false,
    userList: [],
    isEditingUsername: false,
    isEditingPassword: false,
    isFetchingUserList: false,
    editUsernameError: null,
    editPasswordError: null,
    fetchUserListError: null,
  },
  action,
) {
  switch (action.type) {
    case UPDATE_USER:
      return {
        ...state,
        user_id: action.payload.user_id,
        username: action.payload.username,
        email: action.payload.email,
        exp: action.payload.exp,
        isSocial: action.payload.isSocial,
      };
    case EDIT_USER_USERNAME_PENDING:
      return {
        ...state,
        isEditingUsername: true,
      };
    case EDIT_USER_USERNAME_FULFILLED:
      return {
        ...state,
        isEditingUsername: false,
        username: action.payload.username,
        email: action.payload.email,
      };
    case EDIT_USER_USERNAME_REJECTED:
      return {
        ...state,
        isEditingUsername: false,
        editUsernameError: action.payload,
      };
    case EDIT_USER_PASSWORD_PENDING:
      return {
        ...state,
        isEditingPassword: true,
      };
    case EDIT_USER_PASSWORD_FULFILLED:
      return {
        ...state,
        isEditingPassword: false,
      };
    case EDIT_USER_PASSWORD_REJECTED:
      return {
        ...state,
        isEditingPassword: false,
        editPasswordError: action.payload,
      };
    case FETCH_USER_LIST_PENDING:
      return {
        ...state,
        isFetchingUserList: true,
      };
    case FETCH_USER_LIST_FULFILLED:
      return {
        ...state,
        isFetchingUserList: false,
        userList: action.payload,
      };
    case FETCH_USER_LIST_REJECTED:
      return {
        ...state,
        isFetchingUserList: false,
        fetchUserListError: action.payload,
      };
    default:
      return state;
  }
}
