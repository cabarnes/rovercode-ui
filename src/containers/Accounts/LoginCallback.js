import React, { Component } from 'react';
import { Redirect } from 'react-router';
import { Loader } from 'semantic-ui-react';
import { injectIntl } from 'react-intl';
import { connect } from 'react-redux';
import axios from 'axios';
import queryString from 'query-string';
import PropTypes from 'prop-types';
import { withCookies, Cookies } from 'react-cookie';
import jwtDecode from 'jwt-decode';

import { updateValidAuth as actionUpdateValidAuth } from '@/actions/auth';
import { updateUser as actionUpdateUser } from '@/actions/user';

const mapDispatchToProps = (dispatch) => ({
  updateUser: (data) => dispatch(actionUpdateUser({ ...data, isSocial: true })),
  updateValidAuth: (isValidAuth) => dispatch(actionUpdateValidAuth(isValidAuth)),
});

class LoginCallback extends Component {
  constructor(props) {
    super(props);

    this.state = {
      loading: true,
      loginSuccess: false,
      error: null,
    };
  }

  componentDidMount() {
    const {
      cookies, location, match, updateUser, updateValidAuth,
    } = this.props;
    const queryParams = queryString.parse(location.search);

    return axios.post(`/jwt/auth/social/${match.params.service}/login/`, {
      code: queryParams.code,
      state: queryParams.state,
    })
      .then((response) => {
        updateUser(jwtDecode(response.data.token));
        cookies.set('auth_jwt', response.data.token, { path: '/' });
        updateValidAuth(true);
        this.setState({
          loading: false,
          loginSuccess: true,
        });
      })
      .catch((error) => {
        let errorMessage = null;
        if (error.response) {
          errorMessage = error.response.data.non_field_errors;
        }
        this.setState({
          loading: false,
          loginSuccess: false,
          error: errorMessage,
        });
      });
  }

  redirect = () => {
    const { intl } = this.props;
    const { error, loading, loginSuccess } = this.state;

    const loggingIn = intl.formatMessage({
      id: 'app.login_callback.logging_in',
      description: 'Shows the user that the application is logging in',
      defaultMessage: 'Logging in...',
    });

    if (loading) {
      return <Loader content={loggingIn} />;
    }

    if (loginSuccess) {
      return <Redirect to="/" />;
    }

    return (
      <Redirect to={{
        pathname: '/accounts/login',
        state: { callbackError: error },
      }}
      />
    );
  }

  render() {
    return (
      <>
        { this.redirect() }
      </>
    );
  }
}

LoginCallback.propTypes = {
  cookies: PropTypes.instanceOf(Cookies).isRequired,
  location: PropTypes.shape({
    search: PropTypes.string.isRequired,
  }).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      service: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
  updateUser: PropTypes.func.isRequired,
  updateValidAuth: PropTypes.func.isRequired,
  intl: PropTypes.shape({
    formatMessage: PropTypes.func.isRequired,
  }).isRequired,
};

export default withCookies(injectIntl(connect(null, mapDispatchToProps)(LoginCallback)));
