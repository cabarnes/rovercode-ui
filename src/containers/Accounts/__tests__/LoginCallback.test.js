import React from 'react';
import { Redirect } from 'react-router';
import { CircularProgress } from '@material-ui/core';
import { Cookies } from 'react-cookie';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import configureStore from 'redux-mock-store';
import jwtDecode from 'jwt-decode';

import { updateValidAuth } from '@/actions/auth';
import { updateUser } from '@/actions/user';

import LoginCallback from '../LoginCallback';

const mockStore = configureStore();
const store = mockStore();
store.dispatch = jest.fn();

const mock = new MockAdapter(axios);

const location = {
  search: '?state=1234&code=5678',
};
const match = {
  params: {
    service: 'google',
  },
};

test('LoginCallback renders on the page with no errors', () => {
  const cookiesValues = { };
  const cookies = new Cookies(cookiesValues);
  const wrapper = shallowWithIntl(
    <LoginCallback location={location} match={match} store={store} />, {
      context: { cookies },
    },
  ).dive().dive().dive();

  expect(wrapper).toMatchSnapshot();
});

test('LoginCallback displays loader while loading', () => {
  const cookiesValues = { };
  const cookies = new Cookies(cookiesValues);
  const cookiesWrapper = shallowWithIntl(
    <LoginCallback location={location} match={match} store={store} />, {
      context: { cookies },
    },
  );

  const wrapper = cookiesWrapper.dive().dive().dive()
    .dive();

  expect(wrapper.find(CircularProgress).exists()).toBe(true);
  expect(wrapper.find(Redirect).exists()).toBe(false);
});

test('LoginCallback redirects to login after failure', async () => {
  mock.reset();
  mock.onPost('/jwt/auth/social/google/login/').timeout();
  const cookiesValues = { };
  const cookies = new Cookies(cookiesValues);
  const cookiesWrapper = shallowWithIntl(
    <LoginCallback location={location} match={match} store={store} />, {
      context: { cookies },
    },
  );

  const wrapper = cookiesWrapper.dive().dive().dive()
    .dive();

  await wrapper.instance().componentDidMount();
  wrapper.update();

  const redirect = wrapper.find(Redirect);
  expect(redirect.exists()).toBe(true);
  expect(redirect.prop('to')).toStrictEqual({
    pathname: '/accounts/login',
    state: { callbackError: null },
  });
  expect(wrapper.find(CircularProgress).exists()).toBe(false);
  expect(cookies.get('auth_jwt')).toBeUndefined();
});

test('LoginCallback redirects to root after success', async () => {
  const token = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoxLCJ1c2VybmFtZSI6ImFkbWluIiwiZXhwIjoxNTQwMzQzMjIxLCJlbWFpbCI6ImFkbWluQGV4YW1wbGUuY29tIiwib3JpZ19pYXQiOjE1NDAzMzk2MjF9.tumcSSAbKeWXc2QDd7KFR9IGh3PCsyHnCe6JLSszWpc';
  mock.reset();
  mock.onPost('/jwt/auth/social/google/login/').reply(200, {
    token,
  });
  const cookiesValues = { };
  const cookies = new Cookies(cookiesValues);
  const cookiesWrapper = shallowWithIntl(
    <LoginCallback location={location} match={match} store={store} />, {
      context: { cookies },
    },
  );

  const wrapper = cookiesWrapper.dive().dive().dive().dive();

  await wrapper.instance().componentDidMount();
  wrapper.update();

  const redirect = wrapper.find(Redirect);
  expect(redirect.exists()).toBe(true);
  expect(redirect.prop('to')).toBe('/');
  expect(wrapper.find(CircularProgress).exists()).toBe(false);
  expect(cookies.get('auth_jwt')).toBe(token);
  expect(store.dispatch).toHaveBeenCalledWith(updateUser({ ...jwtDecode(token), isSocial: true }));
  expect(store.dispatch).toHaveBeenCalledWith(updateValidAuth(true));
});
