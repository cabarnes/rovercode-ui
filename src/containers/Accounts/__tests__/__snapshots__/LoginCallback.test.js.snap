// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginCallback renders on the page with no errors 1`] = `
<Connect(LoginCallback)
  allCookies={Object {}}
  cookies={
    Cookies {
      "HAS_DOCUMENT_COOKIE": true,
      "changeListeners": Array [
        [Function],
      ],
      "cookies": Object {},
    }
  }
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatDateToParts": [Function],
      "formatDisplayName": [Function],
      "formatHTMLMessage": [Function],
      "formatList": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatNumberToParts": [Function],
      "formatPlural": [Function],
      "formatRelativeTime": [Function],
      "formatTime": [Function],
      "formatTimeToParts": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getDisplayNames": [Function],
        "getListFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralRules": [Function],
        "getRelativeTimeFormat": [Function],
      },
      "locale": "en",
      "messages": Object {},
      "onError": [Function],
      "textComponent": Symbol(react.fragment),
      "timeZone": undefined,
    }
  }
  location={
    Object {
      "search": "?state=1234&code=5678",
    }
  }
  match={
    Object {
      "params": Object {
        "service": "google",
      },
    }
  }
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
/>
`;
